### PR TITLE
Refactored ghetto chem separator

### PR DESCRIPTION
**Pull request by SyncIt21**
## About The Pull Request

This completely reworks how ghetto chem separator works & brings more depth & immersion to it. To understand the new system let's compare it with the current one

**Old System** 
1. You insert a beaker containing the reagents you want to separate.
2. You use a lighter/whatever to light a flame, Flame has _magic & unrealistic_ properties such as
    - It will never heat the beaker above 100k
    - It doesn't burn oxygen or use fuel source to sustain it
    - Distillation rate is a constant 5u/s. It doesn't increase as the solution boils hotter
 
3. The apparatus sorts the reagents alphabetically? & transfers the 1st reagent from sorted list to the target beaker. It does not care if each individual reagent has their boiling points & how to separate them based on that etc. It's just bland overall

**New System**
1. You insert a beaker containing the reagents you want to separate, just lke the old system
2. You use a lighter/whatever to light a flame just like the old system, however flame behaves in a more realistic fashion as follows
    - It will heat the beaker all the way to 1000k if you let it run long enough
    - It now burns oxygen & uses a fuel source(a secondary beaker which you must insert containing flamable reagents like oil, welding fuel etc). When oxygen is the fuel source no O2 gas is required
    - Distillation rate starts at 2u/s & increases with rising temps
    - Intensity of the flame depends on the quality of reagents used as fuel, Oil has best flammable quality & will heat the beaker faster, other reagents mixed in various proportions can produce better quality flames which you can understand by examining the aparatus
    - The flame intensity can be adjusted via the burner knob setting
3. Each reagent now has their _boiling computed_ computed from their mass allowing for realistic interactions, like having to control the temp's just right to separate your required reagent
4. Temps automatically cooldown and you can install a condenser to Facilitate faster cooling

Here's the demo video explaining everything

https://github.com/tgstation/tgstation/assets/110812394/ba591859-3fff-4d06-90f8-024c7266f830

## Changelog
:cl:
refactor: ghetto chem separator has been reworked from scratch. See PR 83275 for details
qol: adds examines & screentips for ghetto chem separator
/:cl:

### CODE REVIEW

**Comment by Metekillot** [comment]
Very cool but ghetto chem is so rarely done as is that I don't understand the motivation behind making one of its tools more difficult and resource intensive to use.

File: tgui/packages/tgui/interfaces/ChemSeparator.tsx [hunk 1]
```diff
@@ -1,159 +1,196 @@
...
       <Window.Content>
-        <Section fill>
-          <LabeledList>
+        <LabeledList>
+          <LabeledList.Item
+            label={
+              <Box
+                style={{
+                  transform: 'translate(20%, -50%)',
+                  width: '57px',
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/ChemSeparator.tsx:34]: You have a lot of pixel measurements going on in this UI, initially not a fan. How does it look when resized? 

Is there a way to write this that doesn't require the spacing shimmed with exact measurements (maybe a table)?

File: tgui/packages/tgui/interfaces/ChemSeparator.tsx [hunk 2]
```diff
@@ -1,159 +1,196 @@
...
+                }}
+              >
+                Flask:
+              </Box>
+            }
+          >
+            <Stack vertical={false} fill>
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/ChemSeparator.tsx:41]: No need for vertical=false here, horizontal is the default for flex/stack

File: tgui/packages/tgui/interfaces/ChemSeparator.tsx (hunk 1 above)
Comment (SyncIt21) [reply tgui/packages/tgui/interfaces/ChemSeparator.tsx:34]: One way would be to use stack inside a stack, (Like the row would be a stack[label would be a stack item and then the 2nd stack item would be the progress bar/button] and each column would be a stack item and then we could adjust the margins to give us spacing).

But yeah it's just not intuitive as you can see from my explanation

**Comment by SyncIt21** [comment]
> ghetto chem is so rarely done as is that I don't understand the motivation behind making one of its tools more difficult and resource intensive to use.

It's because it is "Ghetto" and is rarely done that some depth/immersion is required to make people intrested enough to put in the effort to try it. Even though the new system is more difficult to use, it is a more rewarding experience because of the new moving parts added.

> Also, I believe that reworks != refactors, as was pointed out to me before. This seems like a balance/feature change

This would be balance if I changed some  existing numbers and would be a feature of I added on top of the existing system but i do neither. I am re writing the whole thing by tearing down the existing system and not tweaking/adding something on top of it so the only fitting lable is a refactor.

**Comment by Metekillot** [comment]
An addition of more moving parts and precision does implicitly increase the difficulty of a task; the operative objective in this scenario being distilling chems from other chems with ghetto chem (because for whatever reason you don't have access to a few machine parts and board for a ChemMaster)

Let me reiterate: I do personally think this is cool. But perhaps it might be better served as an addition to the current ChemMaster when it comes to precision chemical separation? People have optimized purity by dumping tons of a base or easy to make reagent before doing their intended reaction to cut out the necessity of managing pH.

Adding this as a next step to extract their reagent could serve a dual purpose of encouraging buffer usage and satisfaction at high purity product creation.

**Comment by ATH1909** [comment]
is the oxygen this required taken from the air, or does it require the specific oxygen reagent? if the latter, that makes this pretty hard to actually do under ghetto chem conditions without a can of canned air or the like, IIRC

**Comment by FlufflesTheDog** [comment]
Not a huge fan of this, seconding the opinion that ghetto chem is already janky and difficult enough. This is a more complicated tool, that requires more work and resources to use, all to get less out of it. This is already a heck of a chore to get your hands on, you'd honestly probably have an easier time getting a regular chem master.

As for the actual content:
You've got K and C mixed up a bit in the PR body and in the UI. The old system heats to 100C not 100K, and your new UI is displaying values in kelvin but presented as celsius.

I believe the old behavior of stopping at a particular temperature is also realistic, even if it only being a single number for any reagent isn't. When boiling a pot of water for example, it doesn't matter how much more heat you dump into it, it's not going above 100C until the water evaporates. The process of evaporation consumes energy, which keeps your input temperature near the lowest boiling point within it.

And why not use an actual knob for the burner knob ui?

**Comment by SyncIt21** [comment]
> is the oxygen this required taken from the air, or does it require the specific oxygen reagent? 

Taken directly from the air, no extra requirements there.

> ghetto chem is already janky and difficult enough. This is a more complicated tool, that requires more work and resources to use, all to get less out of it

Crafting ghetto equipment is the only difficult part. Operating the apparatus themselves is pretty easy, like for the distillation apparatus you only needed to press the start button, wait for like 2 seconds and get your separated reagent. The whole point of Ghetto is "It's supposed to be difficult" which is what this PR brings.

> You've got K and C mixed up a bit in the PR body and in the UI. The old system heats to 100C not 100K, and your new UI is displaying values in kelvin but presented as celsius.

Caught a bug there, I'll fix that.

> I believe the old behavior of stopping at a particular temperature is also realistic, even if it only being a single number for any reagent isn't. When boiling a pot of water for example, it doesn't matter how much more heat you dump into it, it's not going above 100C until the water evaporates

You are talking about the temperature of the reagents present inside which is true, I'm talking about the temperature of the whole reagent holder(beaker) which can be heated indefinitely but transfer it's heat to the reagents inside, so yes this is realistic.

> And why not use an actual knob for the burner knob ui?

I'm goanna implement that is a future PR, and even though we need the burner for crafting, we hold no direct reference to it inside the distillation apparatus so modifying it's UI code is out of scope here.

**Comment by MTandi** [comment]
I like the depth, but I strongly disagree with this:

> Ghetto is "It's supposed to be difficult"

My original intention was to provide some basic and easy to use tool for the case when you can't acquire a chem master. 
It was not aimed for masochistic nerds that want even more complexity on top of the complex fermichem mechanic.

The crafting was already quite tricky but now you need fuel, and you need to wait more for the chems to separate. I wish you made it a bit easier to use, while preserving the details related to the boiling point.

Also, I don't understand by glance how you're supposed to extract the beaker and the burner to refill them. Do you right click on the reagent line?

**Comment by SyncIt21** [comment]
> My original intention was to provide some basic and easy to use tool for the case when you can't acquire a chem master.
> It was not aimed for masochistic nerds that want even more complexity on top of the complex fermichem mechanic.

This PR is me disagreeing with your original intention. I believe anything Ghetto should be clunky and difficult to use for the sake of immersion. Just because it's now a little more difficult doesn't mean it's not fun, some nerds(like me) like the complexity and challenges as it creates a more rewarding experience when you get it done right.

> The crafting was already quite tricky but now you need fuel, and you need to wait more for the chems to separate. I wish you made it a bit easier to use, while preserving the details related to the boiling point.

It's not like I added magic requirements to the apparatus. A fuel source to sustain the flame is what you expect from real life. Apart from that everything else works exactly the same. Considering how easy it is to create fuel like oil I don't think it is too hard of an requirement to have. 
As for crafting, all items can be printed in the autolathe and now you can deconstructed the apparatus with a crowbar allowing you to reuse your materials for other purposes 

> Also, I don't understand by glance how you're supposed to extract the beaker and the burner to refill them. Do you right click on the reagent line?

So you right click to insert a container having fuel and right click with empty hand to remove the same container. Left click is to insert your beaker for distillation

**Comment by FlufflesTheDog** [comment]
> You are talking about the temperature of the reagents present inside which is true, I'm talking about the temperature of the whole reagent holder(beaker) which can be heated indefinitely but transfer it's heat to the reagents inside, so yes this is realistic.

That's not how it's set up to work, though. The displayed temperature is the temperature of the reagents as a whole, you can have two things boiling at the same time with their boiling temperatures 100k apart

> I'm goanna implement that in a future PR, and even though we need the burner for crafting, we hold no direct reference to it inside the distillation apparatus so modifying it's UI code is out of scope here.

I'm talking about the separator, not the literal burner item, which has no ui. You implemented a "knob" ui which is just 5 checkboxes. I'm suggesting using a knob or slider component, assuming it's practical. Something like the music player volume knob
image

And some comments on the actual code:
<details>

- Right clicking to remove the fuel beaker removes both beakers (There should really be a ui interaction for this too, imo)
- Trying to insert a beaker into an occupied slot eats the old beaker
- The ambient light from the flame exists and responds to the knob settings even while unlit
- The sound loop doesn't stop when it finishes boiling
- Taking out the output beaker says you take out the *distilation flask* which is wrong, the flask stays mounted on the device
- You're not able to use subtypes of the listed fuels, eg you can only use ethanol and not any type of booze for fuel, even though you can craft an ethanol burner with any type of ethanol
- Being able to use oxygen as a fuel seems a little weird, with the point of it using oxygen from the air as an oxidizer? Perhaps plasma would be a good addition, too

</details>

**Comment by SyncIt21** [comment]
> That's not how it's set up to work, though. The displayed temperature is the temperature of the reagents as a whole, you can have two things boiling at the same time with their boiling temperatures 100k apart

Let's not sweat the smal details here. The chem heater also heats your beaker upto 1000k and the water inside doesn't boil. Small inconsistencies like this can be overlooked.

> You implemented a "knob" ui which is just 5 checkboxes. I'm suggesting using a knob or slider component, assuming it's practical. Something like the music player volume knob

 Sure I'll see what I can do.

> Details

Thank you. I'll look into it

**Comment by FlufflesTheDog** [comment]
> Let's not sweat the smal details here. The chem heater also heats your beaker upto 1000k and the water inside doesn't boil. Small inconsistencies like this can be overlooked.

You're adding a mechanically detailed feature, for the sake of it being realistic and detailed. This is the sort of thing that needs to be thought out. Inconsistencies can be fine but there's impact to the functionality.
By having the act of boiling not consume energy, you allow multiple reagents to boil at the same time. If two important boiling points are fairly close, that will make it very difficult to separate them. Perhaps that's okay, but it also makes it frustrating to try again if you overheat. Because if you transfer it to the input it will STILL be over the desired temperature and start boiling both chemicals again.

> Yes, that's it's only point. Using plasma would actually make it way too hard to set up

I don't understand. I'll split my original statement into two points because I feel like they got confused.
- You've added /datum/reagent/oxygen as a potential *fuel* source. The separator uses oxygen (the gas) as an oxidizer, burning oxygen in oxygen doesn't really make sense.
- Liquid plasma could work as another possible fuel, being NT's magic energy-packed research material. Of course the other fuels are still an option, but plasma sheets are not too hard to come across in maint.

**Comment by SyncIt21** [comment]
> By having the act of boiling not consume energy, you allow multiple reagents to boil at the same time. If two important boiling points are fairly close, that will make it very difficult to separate them. Perhaps that's okay, but it also makes it frustrating to try again if you overheat

Heat energy is created by burning fuel and O2 so boiling does consume energy? don't know what your referring to. Plus yes it is expected that 2 reagents could have similar boiling points just like real life(google boiling points of ethanol and ethyl)and this is an important limitation of Ghetto distillation which makes you give up and look for an chem master


> You've added /datum/reagent/oxygen as a potential fuel source. The separator uses oxygen (the gas) as an oxidizer, burning oxygen in oxygen doesn't really make sense

Checking if the fuel source contains O2 and deciding if we should burn oxygen gas or not is way too hacky. This would allow you to start a flame in space just because the fuel source contains O2 which isn't realistic at all.

> Liquid plasma could work as another possible fuel, being NT's magic energy-packed research material. Of course the other fuels are still an option, but plasma sheets are not too hard to come across in maint.

Sure I can add plasma as a fuel source. Oxygen reagent is also supported so no worries there.

**Comment by FlufflesTheDog** [comment]
> Heat energy is created by burning fuel and O2 so boiling does consume energy? don't know what your referring to. Plus yes it is expected that 2 reagents could have similar boiling points just like real life(google boiling points of ethanol and ethyl)and this is an important limitation of Ghetto distillation which makes you give up and look for an chem master

Requiring heat and consuming heat are two different things. Again, realistically, a boiling pot of water will never exceed 100c until all the water has been boiled off, because energy is being consumed by the phase change. If you stop providing heat, the boiling stops almost immediately.

Here, you can heat a chemical well past its boiling point. You can turn off the heat, and it will continue to boil, only losing heat and the ambient cooling rate you've set. You can cycle the same chemical through over and over, and it will boil the whole thing over and over again without losing any appreciable heat- the practical impact of this is that you need to find another way to cool the reagents down, if you overshoot the temperature. Whereas realistically it wouldn't be *possible* to overshoot the temperature, but you'd get the same problem if you instead ran the distiller for too long.

> Checking if the fuel source contains O2 and deciding if we should burn oxygen gas or not is way too hacky. This would allow you to start a flame in space just because the fuel source contains O2 which isn't realistic at all.

Not exactly what I'm trying to say. I mean that oxygen itself is not a fuel, it doesn't make much sense to burn it as a fuel. Using oxygen as a fuel in space is not really any different than using oxygen as a fuel on station- you still only have half of the equation.

**Comment by SyncIt21** [comment]
Requiring heat and consuming heat are two different things. Again, realistically, a boiling pot of water will never exceed 100c until all the water has been boiled off, because energy is being consumed by the phase change. If you stop providing heat, the boiling stops almost immediately.

> Here, you can heat a chemical well past its boiling point. You can turn off the heat, and it will continue to boil, only losing heat and the ambient cooling rate you've set. You can cycle the same chemical through over and over, and it will boil the whole thing over and over again without losing any appreciable heat- the practical impact of this is that you need to find another way to cool the reagents down, if you overshoot the temperature. Whereas realistically it wouldn't be possible to overshoot the temperature, but you'd get the same problem if you instead ran the distiller for too long.

I agree with your reasoning but from a code perspective implementing this is just not worth the effort.
The reagent holder has a `chem_temp` and we increase that value. If it exceeds the boiling point of a reagent(which is computed by the `get_boiling_point()` proc) we distill away that reagent. I don't want to make it anymore complicated than this so let's end the discussion of accurate math and physics on this one.

> Using oxygen as a fuel in space is not really any different than using oxygen as a fuel on station- you still only have half of the equation.

Now that I think about what you said it is true that rockets in real life burn liquid O2 and other fuel in space for propulsion.
 Fair enough, if the fuel source contains O2 it won't need to use "Oxygen gas" I'll implement that change 👍

**Comment by FlufflesTheDog** [comment]
First some clarification that I'm not describing an exploit/cheese, but rather a frustrating situation that will inevitably arise. 
Here, I've pre-heated a mixture of chemicals to demonstrate, but you'd end up in a similar situation simply by leaving the burner on too high for too long. 
I add them to the flask, and they *both* immediately start boiling, rather quickly. No appreciable amount of heat is lost in this process, meaning that trying to move them back to the input just repeats the process. You cannot just wait for them to cool off either, because only the flask has the passive cooling which they quickly boil out of. You're locked out of separating these chemicals at all, just because they're too hot.

A possible solution to this is just to cap the temperature of the reagents as a whole to the lowest boiling point within. 
Another would be making it so that the condensate on the output loses some notable amount of its heat after being transferred.

https://github.com/tgstation/tgstation/assets/25628932/c16f7301-c4b7-4979-aba0-cf3897c7dcdf

**Comment by SyncIt21** [comment]
Sorry it took a while for to understand the problem. 

> A possible solution to this is just to cap the temperature of the reagents as a whole to the lowest boiling point within.

While this solution is magic i agree this is the most easiest one to do. I'll make it so that clicking the "Filter" button will automatically adjust the temps back to room value(300K) on the distillation flask(Assume filtration is a magic process that lowers temps). That way reagents don't insta boil.

**Comment by SyncIt21** [comment]
Try testing it now for bugs & suggestions

**Comment by SyncIt21** [comment]
Alright temps now cooldown for the target beaker as well. You can also install condenser to allow for faster cooling but it consumes fuel to do so

**Comment by FlufflesTheDog** [comment]
Certainly a bit better, but I'd personally still prefer the temperature be capped at the lowest boiling point. The energy consumed by the boiling should make it difficult to exceed that temperature.

It takes about 2 minutes to heat a full flask 100 kelvin (using oil). Meaning it might take a whole 12 minutes to separate things with high boiling points. That's frankly somewhat absurd compared to a chem master doing it in two seconds, or even the old separator which could do it in maybe a minute.

And when operating with smaller volumes, you end up with the opposite problem. A single tick on even the lowest setting will increase 15 units by 15-20k, potentially instantly starting to boil multiple things.

There also seems to be a fundamental issue with how boiling points are computed. Just transferring the chemicals around (eg empty and refill the flask) will give you a whole new set of boiling points, meaning every new batch or every time you make a mistake, you need to study the boiling points again to figure out what the new order will be.

For testing, if you want to try yourself, I've been grinding up some ambrosia, enough to fill up the input flask, and trying to separate out the libital and aiuri. It has several "trash" chemicals and two important ones to split off, so it's a good test of how usable it is.

**Comment by SyncIt21** [comment]
> And when operating with smaller volumes, you end up with the opposite problem. A single tick on even the lowest setting will increase 15 units by 15-20k, potentially instantly starting to boil multiple things.

These inconsistencies with heating is actually intentional. The burner imparts a constant amount of energy so smaller volumes will boil faster and larger volumes will boil longer because the heat capacity of reagents(dependent on volumes) vary.

> Meaning it might take a whole 12 minutes to separate things with high boiling points. That's frankly somewhat absurd compared to a chem master

You right. I'll try to cap the max boling points to a lower value and experiment with your test case.

> There also seems to be a fundamental issue with how boiling points are computed. Just transferring the chemicals around (eg empty and refill the flask) will give you a whole new set of boiling points,

This is definitely a bug. I'll fix that.

**Comment by FlufflesTheDog** [comment]
> These inconsistencies with heating is actually intentional. The burner imparts a constant amount of energy so smaller volumes will boil faster and larger volumes will boil longer because the heat capacity of reagents(dependent on volumes) vary.

This sounds like another half-step towards realism. Yes, large volumes should take longer to heat up than small volumes. But it doesn't make sense that you wouldn't be able to set the burner low enough to maintain some amount of control. And larger volumes, for the sake of gameplay, should again probably not take 10 minutes to heat up. 

I've run my test again, and- using plasma, it took me **12 minutes** to fully process 300 units of ground up ambrosia, from ignition to the moment aiuri and libital were isolated in beakers. I assume the boiling points are still random upon world generation, but a breakdown of how it went:
5 minutes before the first chemical reached its boiling point. Absolutely no interaction up to this point, just waiting.
9 minutes before the libital started boiling
and 11:30 minutes before the aiuri started boiling, finishing another 30 seconds after that.
The whole process used about 60 units of liquid plasma.

The point of all that is to say that this is an *incredibly* slow and tedious process, and not one that you can walk away from due to needing to micromanage heating and swap out beakers. For as uninteresting of a reward as you get out of it (separation of chemicals), I can't see people making this as-is outside of the novelty of the mechanics, or not knowing what they're about to get into.

**Comment by SyncIt21** [comment]
> Absolutely no interaction up to this point, just waiting.

That's how real life heating of reagents work. You set the heat on and wait for it to boil, what more interaction do you want from this.

> The point of all that is to say that this is an incredibly slow and tedious process, and not one that you can walk away from due to needing to micromanage heating and swap out beakers

Slow or fast depends on what reagent you are trying to filter out and how much volume of reagents you put in the apparatus.

It's called Ghetto for a reason. I don't expect people to put in 300 units of solution and get amazing results. Maximum 100u and optimal 30u is all the apparatus can handle. Anything more than that and you are expected to get a difficult experience, give up and just look for an chem master.

I won't be tweaking any more numbers. I'm just waiting for this to get merged.

**Comment by Kapu1178** [comment]
Hello @SyncIt21, I have no other way to contact you as you are not present on discord and your repo has no issues page, sorry to derail this PR.

In https://github.com/tgstation/tgstation/pull/80176, you made what I believe to be an error. The old behavior of the pyrosium reaction was to set the reagent holder's temperature to 20, to emulate massive heat loss from an endothermic reaction. Your change makes it *add* 20 arbitraryheatunits instead. Please correct me if I am mistaken on this.

**Comment by FlufflesTheDog** [comment]
No, no, they have a point
image
Expose temperature adjusts temperature, not sets

**Pull request merged**